### ISSUE TITLE
Enhance ADO.NET with ability to create user supplied factories

### ISSUE THREAD

**Issue by veikkoeeva** [dotnet/orleans#4691]
This issue records some ideas from Gitter discussion with regards ADO.NET functionality. This started from an idea to make using "non-official" libraries easier, such as https://mysql-net.github.io/MySqlConnector/, but it appears this library is drop-in compatible with the official one, so for that case this refactoring isn't strictly needed. It may be there are others that are not.

Currently the ADO.NET provider does not surface the options to load the library to the public configuration surface, but instead uses a user-supplied (well-known) ADO.NET invariant to load libraries. It happens at https://github.com/dotnet/orleans/blob/master/src/AdoNet/Shared/Storage/DbConnectionFactory.cs#L26 and the options being set are at https://github.com/dotnet/orleans/blob/master/src/AdoNet/Shared/Storage/DbConstantsStore.cs#L43 and used at https://github.com/dotnet/orleans/blob/master/src/AdoNet/Shared/Storage/RelationalStorage.cs#L274, for instance.

It could be beneficial to:
1. Consider that now that .NET Standard has again included `DbProviderFactory` if it could be used in the future (see at https://github.com/dotnet/corefx/issues/20903),
2. give the user an ability to change some of configuration values in `ConstantsStore` (such as if it's non-blocking or not),
3. give the user the option to supply the path to the connector dll to be loaded,
4. and optionally create an utility function to extract the version of the library so it would become possible to just determine somewhat automatically if the library has certain properties, such as truly non-blocking queries as-per some version.

It's notable these options are there to give some very fine-tuning knobs to certain situations, but aren't strictly necessary.

**Comment by bgrainger** [comment]
> this library is drop-in compatible with the official one

It's compatible but fixes many of the major problems with Connector/NET; for example, async operations are 100% supported. Thus `isSynchronousAdoNetImplementation` (and possibly `supportsCommandCancellation`) at https://github.com/dotnet/orleans/blob/f38a6f91129a8f99fd7ace5d748042e982628122/src/AdoNet/Shared/Storage/DbConstantsStore.cs#L30-L37 are incorrect for MySqlConnector.

**Comment by veikkoeeva** [comment]
@bgrainger You are right. These are something to be surfaced on the configuration interface with the usual ASP.NET style. You may have noticed, but I write for the general benefit, one difference is that Orleans is split among functionality and not by vendor, so there aren't function variants with those `SqlServer` etc. and instead the connector library is loaded using the well known ADO.NET invariant string.

Also to note is that the current method of having the well known ADO.NET invariant separately as it is done dates partially back to Orleans 1 (and earlier) when due to how the configuration worked it had to be piggy-backed though the system in a connection string (the system didn't know which library to load otherwise in .NET Full and using `DbProviderFactory` couldn't be used). When the configuration was overhauled in the 2.n series, the old system was improved and brought into this configuration surface. We have new options now.

Now that [DbProviderFactory](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/dbproviderfactories) is back in the newest .NET Core (Standard too?), it's also an option to consider when overhauling the configuration.
